File: buffer/cpu/src/jvmMain/java/com/wsr/cpu/elementwise/operation/times/JTimes.java
package com.wsr.cpu.elementwise.operation.times;

import java.nio.ByteBuffer;

public class JTimes {
    // 0次元
    public native void timesD0ToD1(float x, ByteBuffer y, ByteBuffer result);

    // 1次元
    public native void timesD1ToD0(ByteBuffer x, float y, ByteBuffer result);

    public native void timesD1ToD1(ByteBuffer x, ByteBuffer y, ByteBuffer result);

    public native void timesD1ToD2(ByteBuffer x, ByteBuffer y, int yi, int yj, int axis, ByteBuffer result);

    public native void timesD1ToD3(ByteBuffer x, ByteBuffer y, int yi, int yj, int yk, int axis, ByteBuffer result);

    // 2次元
    public native void timesD2ToD1(ByteBuffer x, int xi, int xj, ByteBuffer y, int axis, ByteBuffer result);

    public native void timesD2ToD3(
            ByteBuffer x,
            int xi,
            int xj,
            ByteBuffer y,
            int yi,
            int yj,
            int yk,
            int axis1,
            int axis2,
            ByteBuffer result
    );

    // 3次元
    public native void timesD3ToD1(ByteBuffer x, int xi, int xj, int xk, ByteBuffer y, int axis, ByteBuffer result);

    public native void timesD3ToD2(
            ByteBuffer x,
            int xi,
            int xj,
            int xk,
            ByteBuffer y,
            int yi,
            int yj,
            int axis1,
            int axis2,
            ByteBuffer result
    );

    public native void timesD3ToD4(
            ByteBuffer x,
            int xi,
            int xj,
            int xk,
            ByteBuffer y,
            int yi,
            int yj,
            int yk,
            int yl,
            int axis1,
            int axis2,
            int axis3,
            ByteBuffer result
    );

    // 4次元
    public native void timesD4ToD1(ByteBuffer x, int xi, int xj, int xk, int xl, ByteBuffer y, int axis, ByteBuffer result);

    public native void timesD4ToD2(
            ByteBuffer x,
            int xi,
            int xj,
            int xk,
            int xl,
            ByteBuffer y,
            int yi,
            int yj,
            int axis1,
            int axis2,
            ByteBuffer result
    );

    public native void timesD4ToD3(
            ByteBuffer x,
            int xi,
            int xj,
            int xk,
            int xl,
            ByteBuffer y,
            int yi,
            int yj,
            int yk,
            int axis1,
            int axis2,
            int axis3,
            ByteBuffer result
    );
}

File: buffer/cpu/src/jvmMain/kotlin/com/wsr/cpu/elementwise/operation/times/JTimes.kt
package com.wsr.cpu.elementwise.operation.times

import java.nio.ByteBuffer

class JTimes {
    external fun timesD0ToD1(x: Float, y: ByteBuffer, result: ByteBuffer)

    external fun timesD1ToD0(x: ByteBuffer, y: Float, result: ByteBuffer)
    external fun timesD1ToD1(x: ByteBuffer, y: ByteBuffer, result: ByteBuffer)
    external fun timesD1ToD2(x: ByteBuffer, y: ByteBuffer, yi: Int, yj: Int, axis: Int, result: ByteBuffer)
    external fun timesD1ToD3(x: ByteBuffer, y: ByteBuffer, yi: Int, yj: Int, yk: Int, axis: Int, result: ByteBuffer)

    external fun timesD2ToD1(x: ByteBuffer, xi: Int, xj: Int, y: ByteBuffer, axis: Int, result: ByteBuffer)
    external fun timesD2ToD3(
        x: ByteBuffer,
        xi: Int,
        xj: Int,
        y: ByteBuffer,
        yi: Int,
        yj: Int,
        yk: Int,
        axis1: Int,
        axis2: Int,
        result: ByteBuffer,
    )

    external fun timesD3ToD1(x: ByteBuffer, xi: Int, xj: Int, xk: Int, y: ByteBuffer, axis: Int, result: ByteBuffer)
    external fun timesD3ToD2(
        x: ByteBuffer,
        xi: Int,
        xj: Int,
        xk: Int,
        y: ByteBuffer,
        yi: Int,
        yj: Int,
        axis1: Int,
        axis2: Int,
        result: ByteBuffer,
    )
    external fun timesD3ToD4(
        x: ByteBuffer,
        xi: Int,
        xj: Int,
        xk: Int,
        y: ByteBuffer,
        yi: Int,
        yj: Int,
        yk: Int,
        yl: Int,
        axis1: Int,
        axis2: Int,
        axis3: Int,
        result: ByteBuffer,
    )

    external fun timesD4ToD1(
        x: ByteBuffer,
        xi: Int,
        xj: Int,
        xk: Int,
        xl: Int,
        y: ByteBuffer,
        axis: Int,
        result: ByteBuffer,
    )
    external fun timesD4ToD2(
        x: ByteBuffer,
        xi: Int,
        xj: Int,
        xk: Int,
        xl: Int,
        y: ByteBuffer,
        yi: Int,
        yj: Int,
        axis1: Int,
        axis2: Int,
        result: ByteBuffer,
    )
    external fun timesD4ToD3(
        x: ByteBuffer,
        xi: Int,
        xj: Int,
        xk: Int,
        xl: Int,
        y: ByteBuffer,
        yi: Int,
        yj: Int,
        yk: Int,
        axis1: Int,
        axis2: Int,
        axis3: Int,
        result: ByteBuffer,
    )
}
